[Proposal] Dynamically add & remove items from the search list

nfvs opened the pull request:
Sometimes it's be useful to be able to add/remove items to the search list without recreating the Fuse() object. This is especially true when using Fuse to search inside a Backbone.js collection, since there are multiple add/remove events and it makes sense to add/remove those from the search list without recreating the Fuse object at every step.

Admittedly the `Fuse()` constructor is rather small so perhaps the overhead of doing is negligible.

This PR is obviously not ready to be merged as is, if you think this is a good addition I will add comments and tests, as well as a way to remove more than just strings.

krisk commented:
I like this. Please add the relevant tests, then we can definitely merge.

krisk commented on src/fuse.js:
```diff
@@ -336,6 +336,19 @@
+    if (!Utils.isArray(to_add)) {
+      to_add = [to_add];
+    }
+    this.list = this.list.concat(to_add);
```
Perhaps do this instead:

`Array.prototype.push.apply(this.list, Utils.isArray(to_add) ? to_add : [to_add])`

krisk commented on src/fuse.js:
```diff
@@ -336,6 +336,19 @@
+    this.list = this.list.concat(to_add);
+  };
+
+  Fuse.prototype.remove = function(to_remove) {
```
Shouldn't this remove by index instead? If it's by string, it won't cover all the cases.

nfvs replied:
That's true, and it's one of the reasons I think this is definitely not ready to be merged. But removing by index is not very useful, at least when used to keep it in sync with a backbone collection, since the order will never be the same.

I was thinking of having `to_remove` be a string for the simplest case where entries are strings, or an object like `{key: value}`, making it work like Backbone's [findWhere](http://backbonejs.org/#Collection-findWhere) where it removes objects based on the passed attributes. Perhaps it can also be a function that would return the element to be removed, making it even more flexible.

nfvs commented:
I've just added a few removal options, if you agree with them I'll implement the tests.

- If the list contains strings:
  - remove by index
  - remove the first matched value
- If the list contains objects:
  - remove by id (only if an id key was specified)

nfvs commented:
Bump

krisk commented on src/fuse.js:
```diff
@@ -337,6 +337,67 @@
+  Fuse.prototype.remove = function(to_remove) {
+    to_remove = Utils.isArray(to_remove) ? to_remove : [to_remove];
+    for (var i = 0, len = to_remove.length; i < len; i++) {
+      Fuse.prototype.remove.apply(to_remove[i]);
```
Did you mean to call `_remove` instead?

krisk commented on src/fuse.js:
```diff
@@ -337,6 +337,67 @@
+   * Dynamically remove a single item from the list
+   * @private
+   */
+  Fuse.prototype._remove = function(to_remove) {
```
Perhaps consider moving this to a function outside of `Fuse.prototype`, since you'll probably be calling with with `apply` anyway.

nfvs replied:
How do you mean? Don't we need it inside the prototype so we can access `this.list`? There are plenty of private functions in `Fuse.prototype` anyway.

RobertHerhold commented:
Is this PR still active? I'd like to see this change implemented

nfvs commented:
I've rebased from master, fixed your suggestions and added unit tests for all cases (adding elements, removing by index/string when using a list of strings, by id when using a list of objects).

oyeanuj commented:
@krisk Is this still on your radar to merge?

Vpr99 commented:
+1

renatodeleao commented:
Same usecase but using `fuse.js` as filterFunction for a vue.js component that returns dynamic (updatable) lists.

Ktar5 commented:
I'm not sure why this was closed, it doesn't seem to be implemented, and if it is, it's nowhere to be seen in the docs?
@krisk Would it be possible to get a quick update on the situation and if it isn't implemented, a suggestion for a workaround to implement this? 

Thanks!

krisk commented:
See #412. Will close this :)
